sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java

```diff
@@ -126,27 +133,134 @@
-                response -> {
-                    logger.verbose("Token refreshed: {}", response);
-                    hasAuthorized.set(true);
-                },
-                error -> {
-                    logger.info("clientId[{}], path[{}], linkName[{}] - tokenRenewalFailure[{}]",
-                        handler.getConnectionId(), this.entityPath, getLinkName(), error.getMessage());
-                    hasAuthorized.set(false);
-                }, () -> hasAuthorized.set(false))
+            () -> {
+                if (tokenManager != null) {
+                    this.tokenManager.getAuthorizationResults().subscribe(
+                        response -> {
+                            logger.verbose("Token refreshed: {}", response);
+                            hasAuthorized.set(true);
+                        },
+                        error -> {
+                            logger.info("clientId[{}], path[{}], linkName[{}] - tokenRenewalFailure[{}]",
+                                handler.getConnectionId(), this.entityPath, getLinkName(), error.getMessage());
+                            hasAuthorized.set(false);
+                        }, () -> hasAuthorized.set(false));
+                }
+            }
         );
     }
 
+    /**
+     * Create AMQP {@link Message} for completing the transaction.
+     *
+     * @param transaction to be used for completing the transaction.
+     * @param isCommit for commiting or rollback
+     * @return created AMQP {@link Message}.
+     */
+    private Message getCompleteTransactionMessage(AmqpTransaction transaction, boolean isCommit) {
+        Message message = Proton.message();
+        Discharge discharge = new Discharge();
+        discharge.setFail(!isCommit);
+        discharge.setTxnId(new Binary(transaction.getTransactionId().array()));
+        message.setBody(new AmqpValue(discharge));
+        return message;
+    }
+
+    /**
+     * Create AMQP {@link Message} for creating the transaction.
+     *
+     * @return created AMQP {@link Message}.
+     */
+    private Message getCreateTransactionMessage() {
+        Message message = Proton.message();
+        Declare declare = new Declare();
+        message.setBody(new AmqpValue(declare));
+        return message;
+    }
+
+    /**
+     * Completes the transaction. All the work in this transaction will either rollback or committed as one unit of
+     * work.
+     *
+     * @param transaction that needs to be completed.
+     * @param isCommit true for commit and false to rollback this transaction.
+     *
+     * @return a completable {@link Mono} which represent {@link DeliveryState}.
+     */
+    public Mono<Void> completeTransaction(AmqpTransaction transaction, boolean isCommit) {
+        Mono<DeliveryState> completeTransaction = Mono.create(sink -> sendTransaction(
+            getCompleteTransactionMessage(transaction, isCommit), sink));
+        if (hasConnected.get()) {
```

Comment:
This hasConnected.get() or retryUtil occurs a lot. We make this into a private method that invokes some Publisher<T>.